feat(ci): create bitcoid docker image and setup regtest mode

vdovhanych opened the pull request:
This pr is the first step to introducing new functionality to tenv. This adds bitcoin core running in regtest mode and also our blocbook service that is connected to there. 

This is still a work in progress. Just putting it here so anyone interested can track the changes.

grdddj commented:
Hello, I am afraid I am not able to validate the correctness of these changes :)

My only comment would be on the branch you started on - the first two commits of mine should not be used, we decided not to touch this SLEEP interval. I would consider rebasing on master - but not needed now, only at the end

vdovhanych commented:
Switching this from the draft. Some key points to this: 

- For now, the building of the image is relying on my fork of blockbook where I put binaries as a release (we would like to do this for the blockbook main repo and have it fetch from there) 
- Image should always be used as`:latest `
- added this image to tenv image repository on GitLab with a new build and push job.
- also added dependency proxy to tenv GitLab (was running into rate limiting on docker hub) 
- we need to update the readme with how to use it and what it's for.
 If you pull the image and run it with compose, you will get our backend in regtest mode, where you can send commands using bitcoin-cli.

tsusanka commented on docker/bitcoin-regtest/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM debian:buster-slim
```
You will be the one who maintains it so keep debian, but you could install blockbook and bitcoin via Nix more easily :stuck_out_tongue:.

vdovhanych replied:
As our backends are Debian based I wanted to have a similar setup to that.

tsusanka replied:
That seems reasonable!

tsusanka reviewed (commented):
- Where are the configs for the bitcoind and blockbook or none are needed?
- Thinking about terminology: we'll call the main image `trezor-user-env` and now the regtest `trezor-user-env-regtest` which kind of doesn't say what the first one contains. But that's probably not something we can decide now, but we should reevaluate in case we would add another docker image.
- Could you document this in the README? 1. Add this to the `Basic terminology`. I would even describe that we have two images (and one that contains the four items there).

> For now, the building of the image is relying on my fork of blockbook where I put binaries as a release (we would like to do this for the blockbook main repo and have it fetch from there)

Could you make an issue please? Either here on in blockbook.

tsusanka commented:
Also please document that the default RPC user and password is `rpc/rpc`.

tsusanka commented on docker/compose.yml:
```diff
@@ -16,3 +16,12 @@ services:
+    ports:
+      - "19021:19021" # internal blockbook port
+      - "19121:19121" # public blockbook port for website dashboard
+      - "18021:18021" # bitcoin-d rpc port
```
As per `bitcoin-cli` help, regtest has port 18443 as default. We should follow that.

vdovhanych replied:
yes, but we are trying to follow our blocbook project ports [table here](https://github.com/trezor/blockbook/blob/51d10bccd207e61eb0ed32702635b6f3f8ea33de/docs/ports.md) Let. me know if it's ok or we should change it.

tsusanka replied:
Ah okay then :shrug:. If it is documented there let's keep it.

vdovhanych commented on docker/bitcoin-regtest/entrypoint.sh:
```diff
@@ -0,0 +1,18 @@
+
+# start bitcoin backend
+
+# change to working directory
```
Yeah, I will get rid of it, it's pretty self-explanatory.

tsusanka replied:
Looks good

vdovhanych commented on ci/build.yml:
```diff
@@ -31,3 +32,23 @@ build and push test image:
     - docker push $CONTAINER_NAME:test
+
+# regtest docker images build
+.build_regtest: &build_regtest
```
I did it this way in case we decide to do testing image the same way as for the base image. But you're eighth it's not necessary and blockbook is released once a year so that's fine to not have testing image.

tsusanka replied:
Aaah I see :+1:. Well I think it's fine then, I was just wondering. Up to you then!

vdovhanych commented on docker/compose.yml:
```diff
@@ -16,3 +16,12 @@ services:
+      - "19021:19021" # internal blockbook port
+      - "19121:19121" # public blockbook port for website dashboard
+      - "18021:18021" # bitcoin-d rpc port
+      - "48321:48321" # message queue
```
This backend service ports (zmq) not really sure what it's used for exactly but it's part of blockbooks documentation.

tsusanka replied:
I do not think we need to expose these. @martinboehm?

vdovhanych commented:
> - Where are the configs for the bitcoind and blockbook or none are needed?
> 
> - Thinking about terminology: we'll call the main image `trezor-user-env` and now the regtest `trezor-user-env-regtest` which kind of doesn't say what the first one contains. But that's probably not something we can decide now, but we should reevaluate in case we would add another docker image.
> 
> - Could you document this in the README? 1. Add this to the `Basic terminology`. I would even describe that we have two images (and one that contains the four items there).
> 
> 
> 
> > For now, the building of the image is relying on my fork of blockbook where I put binaries as a release (we would like to do this for the blockbook main repo and have it fetch from there)
> 
> 
> 
> Could you make an issue please? Either here on in blockbook.

- config for blockbook and bitcoind are a part of blockbook master now we merged the changes where Bitcoin in retreat mode was [introduced](https://github.com/trezor/blockbook/pull/648)

- I think that it wouldn't hurt to maybe redo the naming all together and rename it based on that decision. 

- yes, will add info to readme and terminology. 

- yes I will open an issue regarding this.

tsusanka commented:
> config for blockbook and bitcoind are a part of blockbook master now we merged the changes where Bitcoin in retreat mode was introduced

Smart! Thanks.

tsusanka commented:
So i think this waits for some documention and we can merge then!

grdddj commented on docker/bitcoin-regtest/entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+set -e
+
+# start bitcoin backend
+cd /opt/coins/nodes/bitcoin_regtest
```
I am noticing that quite late, but are these two `cd`'s necessary? Seems all the filenames we use are absolute, so our location does not matter.

If they are really redundant, we can choose if we remove them or make the locations below relative to them (I would vote for the latter, as the blockbook command is very long).

vdovhanych replied:
This is needed for the working directory of blocbook and bitcoind as it expects to be launched with systemd, which sets this directory. This is the solution not to use systemd in docker as that is something we do not want to have.

grdddj replied:
Thanks for the explanation. Cannot still the commands below be made shorter by making them relative, when we are already in that directory?

`./bin/bitcoind -datadir=./backend -conf=./bitcoin_regtest.conf &`

Applies maybe more for the second very long command

vdovhanych replied:
It's not all in the same directory as the working one. I could shorten it a bit but not that much.

grdddj replied:
I see, both commands contain `/opt/coins/data/` which we are not in. We can certainly live without it, but I would welcome the shortening, if it does not have any disadvantages

vdovhanych replied:
Tried it with relative commands but it wasn't working.

tsusanka replied:
@grdddj's points are good but I think we'll survive if it is causing troubles. Resolving.

grdddj commented on README.md:
```diff
@@ -82,7 +84,13 @@ For a future use you can omit the second step and run `up` (the third step) dire
 
 Currently not supported, but should work using docker as well.
+### Bitcoin regtest
+Optionally you can run bitcoin regtest backend with trezor-user-env. This will run another docker container that contains bitcoind in regtest mode and our blocbook backend connected to that bitcoind in regtest mode.
```
Typo - `blocbook`

grdddj replied:
On the optionallity - `controller` is currently always connecting to the regtest at the beginning, so we should have error handling in place when the regtest would not be running

grdddj commented on README.md:
```diff
@@ -82,7 +84,13 @@ For a future use you can omit the second step and run `up` (the third step) dire
 
+1. Clone this repo `git clone git@github.com:trezor/trezor-user-env.git`.
+2. Download the latest docker build: `docker-compose -f ./docker/compose.yml pull trezor-user-env-regtest`
+3. Run it: `docker-compose -f ./docker/compose.yml up trezor-user-env-regtest`
```
We could also have the section with commands running both containers at the same time - but that would depend on the usage, if regular users would be using the regtest functionality often

vdovhanych replied:
Ok let's do it that way then

tsusanka commented:
Let's finish this first and tackle https://github.com/trezor/trezor-user-env/issues/102 after this is merged.

vdovhanych commented on docker/bitcoin-regtest/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+  && apt-get clean \
+  && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+RUN wget https://github.com/vdovhanych/blockbook/releases/download/v0.3.5/backend-bitcoin-regtest_0.21.1-satoshilabs-1_amd64.deb \
```
that is where the binaries are built and deployed for now, will change that to trezor/blockbook with this [issue](https://github.com/satoshilabs/devops/issues/45)

tsusanka replied:
Ah I remember discussing that. Apologies. Maybe add a comment for other ignorants like me?

vdovhanych replied:
done

vdovhanych commented:
addressed  the comments in [commit](https://github.com/trezor/trezor-user-env/pull/98/commits/470410fa5bb0f9002ca6d51acaeebd0386d0c6e7)

tsusanka commented on README.md:
```diff
@@ -28,6 +28,9 @@ This environment is meant to support Trezor development - both Firmware and Suit
   - Bitcoind with [blockbook](https://github.com/trezor/blockbook) running in regtest mode
+  - Default credentials for bitcoin backend
+    - user and password: `rpc/rpc` and that you can run
+  - test the connection with `bitcoin-cli -rpcport=18021 -rpcuser=rpc -rpcpassword=rpc -getinfo`
```
Capital `t`. I know it is nit-picking but let's have nice readmes!
You can also add "You can test.."

vdovhanych replied:
done

tsusanka commented on README.md:
```diff
@@ -64,7 +67,7 @@ For a future use you can omit the second step and run `up` (the third step) dire
 - [Docker](https://docs.docker.com/docker-for-mac/install/) (to run NixOS in a container)
 - [XQuartz](https://www.xquartz.org/) (to share your screen with Docker)
-
+- Intel Mac device (ARM Mac's are not supported for now)
```
Let's put this under the `### MacOS` directly without `-`. And maybe in _italic_.

vdovhanych replied:
done

tsusanka reviewed (changes requested):
Few more nits. Feel free to rebase and force-push then.

tsusanka commented:
You forgot https://github.com/trezor/trezor-user-env/pull/98/files#r687022814 ..
